### ISSUE TITLE
[Feature Request] An option to forbidden packages to upgrade from a attested version to a unattested version

### ISSUE THREAD

**Issue by SukkaW** [pnpm/pnpm#8889]
### Contribution

- [ ] I'd be willing to implement this feature ([contributing guide](https://github.com/pnpm/pnpm/blob/main/CONTRIBUTING.md))

### Describe the user story

Rspack recently encountered a token theft attack where it seems that the npm classic token they used on CI is leaked and attackers are using this to publish a malicious version of the package. See https://github.com/web-infra-dev/rspack/issues/8767

Rspack uses npm provenance when publishing new versions using the CI. When the attacker is trying to publish the malicious version of the package, he/she/it doesn't have access to the CI, so the attacker publishes the package locally, resulting in this version not having attestation.

----

**Update**

On Aug 28, 2025, nx's publish token was compromised. Since the attackers have no access to the CI/CD infra, the malicious versions were published without provenance.

This once again shows how important this feature is.

### Describe the solution you'd like

Forbidding an already attested package from upgrading to an unattested version can reduce this kind of risk since the attackers need to have access to both the npm and the GitHub in order to publish an attested malicious version.

### Describe the drawbacks of your solution

It is possible that a maintainer deliberately disables npm provenance.

**Comment by zkochan** [comment]
I guess we could check if there are any attested versions of the package in the metadata. If there are such versions, then we could disable the installation of non-attested versions.

However, this would mean that we need to loop over all the versions in the metadata to search for the attested field. We need to check how it will influence performance.

**Comment by zkochan** [comment]
There was another issue related to this that suggested a different approach, disabling running the postinstall script: https://github.com/pnpm/pnpm/issues/8891

We do have currently a way to declare an allowlist of packages that are allowed to be built (related setting: [pnpm.onlyBuiltDependencies](https://pnpm.io/package_json#pnpmonlybuiltdependencieshttps://pnpm.io/package_json#pnpmonlybuiltdependencies)).

In this case the dependency did not have a postinstall script prior to the attack. So, if someone is using an allowlist, the postinstall would not work. This approach would not work though if the attacked dependency used to have a postinstall prior to the attack, as in that case it would already be in the allowlist.

Maybe we could make pnpm safer by requiring users to declare "trusted dependencies" that are allowed to be built.

**Comment by zkochan** [comment]
> Most people don't know how to audit if newer versions are safe or not and most of them would simply add new versions (including malicious ones) to the trusted list to silence the warning.

Why would they add the package to the trusted list if it works without the lifecycle script execution? Also, pnpm is not printing any warnings about blocked scripts. Only an info message is printed.

I am not saying this should replace your suggestion but it should be done regardless.

> We only check that when a package's version becomes newer (through an update), we then check the current version's attestation status and the incoming version's attestation status. This way we don't need to loop through entire package metadata, only the current version info and incoming version info.

I don't understand why a distinction should be introduced between update and adding a new dependency. Why this check shouldn't make adding a new dependency safer as well?

**Comment by SukkaW** [comment]
> Why would they add the package to the trusted list if it works without the lifecycle script execution?

Let's say a package using the notorious `node-gyp` (which means this package would rely on `install` or `postinstall` scripts to work) and people would add it to the trusted list. Now the package is compromised, people would still most likely update the version in the trusted list without any audition, and they would simply think "Why do I need to update this list every time I need to upgrade a dependency, that's so frustrating!"

> I don't understand why a distinction should be introduced between update and adding a new dependency. Why this check shouldn't make adding a new dependency safer as well?

Let's say a package that has many popular versions published before provenance is even introduced by npm. What happens if people want to install those old versions (the dep update is scheduled but has yet to happen) even though it is not attested?

**Comment by zkochan** [comment]
I don't claim that blocking scripts will solve all the issues but it will add one more layer of security.

> Let's say a package that has many popular versions published before provenance is even introduced by npm. What happens if people want to add those old versions (e.g. newer versions contain breaking changes that are very difficult to migrate) even though it is not attested?

The versions published before the first version with provenance would be allowed.

**Comment by zkochan** [comment]
Does attestation only make sense in prebuilt packages? Regular js packages are publicly verifiable anyway.

**Comment by SukkaW** [comment]
> Does attestation only make sense in prebuilt packages? Regular js packages are publicly verifiable anyway.

One of the npm provenance requirements is that the "verified" package must be published from a supported building infra (currently GitHub Actions and GitLab CI/CD are supported). This means for an attacker to publish a "verified" malicious version, he/she/it must also have access to the CI. This adds an extra layer of defense. In the rspack incident, the attacker doesn't have access to the GitHub Actions / doesn't know how npm provenance works, hence the malicious version was published from his/her/its own computer, hence the not attested.

**Comment by tats-u** [comment]
At least pnpm should warn users and confirm how serious they are when they try to upgrade from a version with a provenance to one without one.

**Comment by sxzz** [comment]
FYI: taze is adding this feature: https://github.com/antfu-collective/taze/pull/198


<img width="840" height="124" alt="image" src="https://github.com/user-attachments/assets/1773937c-9667-4d8e-ae6e-79b3832d8044" />

**Comment by SuperchupuDev** [comment]
just fyi nx's publish token was compromised yesterday, leading to malicious versions without provenance being published. nx normally uses provenance, so the functionality proposed in this issue would have made pnpm users much less affected

**Comment by SukkaW** [comment]
> just fyi nx's publish token was compromised yesterday, leading to malicious versions without provenance being published. nx normally uses provenance, so the functionality proposed in this issue would have made pnpm users much less affected. something like what taze implemented would be great

@SuperchupuDev Just updated my issue description to include the nx case as well!

**Comment by dominikg** [comment]
Do you have to crawl all versions? I think you just have to check the currently installed version. If that has provenance and the version you are requesting to be installed does not, it should be blocked by default and require explicit approval by the user either through a global flag that disables this check or an allow list

**Comment by dominikg** [comment]
While preventing updates to freshly released packages can help keep you safe for some time, if everyone did that, you'd end up in the same situation just 3 days later. Both serve a different case in my opinion. 

Delay reduces impact because a few brave early adopters might detect it and it gets fixed before it reaches a majority (for that, delay would have to be the default setting, i'd put that on the list for next major)

Throwing an error for loss of provenance and possibly some relevant provenance metadata (repo name, workflow file) not only reduces impact but actively alerts people of potentially malicious verisons, leading to them being investigated and removed faster.

Ideally the registry itself would offer better tools and settings to prevent malicious versions from getting published in the first place, but package managers doing validation before install or running scripts is still going to help.

**Comment by zkochan** [comment]
> While preventing updates to freshly released packages can help keep you safe for some time, if everyone did that, you'd end up in the same situation just 3 days later. Both serve a different case in my opinion.

That's not the case as there are multiple companies continuously scanning the registry.

I am not saying provenance is not useful. We will probably implement this too. However, nothing prevents the author/owner of a package to publish malware to their own package using provenance...

**Comment by tats-u** [comment]
> it won't work on "pnpm add nx". 

I think we should split it into another issue.

> We have released an even stronger check IMO using date: https://pnpm.io/blog/releases/10.16#new-setting-for-delayed-dependency-updates

Thank you for such a nice feature, but suppose the following situation:

- 1.0.2 (same as 1.0.0 except for version number): half day ago
- 1.0.1 (malicious, deprecated): 1 and half days ago
- 1.0.0: 1 week ago

I wonder which version is chosen when `minimumReleaseAge` is 1440 (1 day). Does it consider version deprecation?

**Comment by samchungy** [comment]
This action seems to have some logic we could copy? https://github.com/danielroe/provenance-action

**Comment by beeequeue** [comment]
> I believe it should work in all cases, not only on update.

Then it could be configurable how many previous versions to check for provenance specifically when installing a new package, defaulting to 1 or 2?
When upgrading it would just check the current and next version

> nothing prevents the author/owner of a package to publish malware to their own package using provenance...

That's a different attack vector than the one this idea is trying to prevent though, I don't think that the fact there are other ways to publish malware should count against adding this feature :^)

**Comment by Martinspire** [comment]
Very cool to see. Is there a way to verify the current provenance? Would be neat to see what a package has before installation (it would increase my preference of deciding what package to install if I have multiple options)

**Comment by ai** [comment]
Do we have any allow-list for this feature (thanks for this feature it is amazing)?

Right now we have a report about `chokidar@4.0.3` which was released 1 year ago and looks good.

```
 ERR_PNPM_TRUST_DOWNGRADE  High-risk trust downgrade for "chokidar@4.0.3" (possible package takeover)
```

**Comment by zkochan** [comment]
> Very cool to see. Is there a way to verify the current provenance? Would be neat to see what a package has before installation (it would increase my preference of deciding what package to install if I have multiple options)

Versions published with provenance are marked with a green checkmark in the registry

<img width="142" height="92" alt="Image" src="https://github.com/user-attachments/assets/c5950135-a57b-48a7-b9fe-a083010ed810" />

> Do we have any allow-list for this feature (thanks for this feature it is amazing)?

Not yet. We can implement it, if needed. Here's one: https://github.com/pnpm/pnpm/issues/10164

**Comment by ai** [comment]
> Not yet. We can implement it, if needed.

We created an issue on chokidar https://github.com/paulmillr/chokidar/issues/1440

But we need some solution to deal with dependencies reverting provenance.

As m,aintainer I would say that there are some cases when you revert provenance for a reason. For instance, if you move from CI release to yubikey based 2FA (which is much secure than CI).

**Comment by tats-u** [comment]
Waiting for https://github.com/SchemaStore/schemastore/issues/5119 (pnpm-workspace.yaml)

**Comment by Timeless0911** [comment]
`pnpm dedupe` will throw an `ERR_PNPM_TRUST_DOWNGRADE` error when checking dependencies but `pnpm install` not, is this a bug?

https://github.com/web-infra-dev/rslib/actions/runs/19322701784/job/55267109693?pr=1333
